### ISSUE TITLE
Create a third way

### ISSUE THREAD

**Issue by iangow** [iangow/wrds2pg#33]
There are two kinds of repositories that the recently updated `wrds2pg` supports.

 - A PostgreSQL database with data organised into schemas like those used by WRDS.
 - A parquet file--based repository organised into schemas like those used by WRDS.

In each case, `wrds2pg` allows for updates contingent on the WRDS data being newer than the local data. @EvanJo has made pull requests that suggest a third alternative: a repository of CSV files.

My inference is that there is some support for contingent updates using the `mtime` of files to support. I wonder how robust this is. (Does it work on Linux, MacOS, and Windows? Does it work across time zones?)

Another issue with CSVs is that the parquet and PostgreSQL options obtain (and retain) information about data types from WRDS. Does it make sense to do something similar for CSVs? (I would say that for the most part, WRDS data is not complex and it may not be necessary. For example, date-times on WRDS need some care, but I'm not sure things are even there with the PostgreSQL/parquet options.)

I think that the CSV repository needs to live in a different place from the parquet option if a user wants to have (parts of) both. Maybe `CSV_DIR` instead of `DATA_DIR`.

Also, there seem to be two `wrds_to_csv()` functions floating around. One is to support `wrds_to_parquet()` and one to support this third way. Perhaps the latter could become `wrds_update_csv()`. This could embed the check `mtime` code and call  `wrds_to_csv()` to get the data.

**Comment by EvanJo** [comment]
I can take a stab. mtime is in epoch time so that should be fine. The sas meta-data seems to be on central time.

Rather than converting to a time stamp, which could get messed up depending on the user's locale settings,  I think it'd be best to just convert the sas time stamp to the proper UTC time.

**Comment by EvanJo** [comment]
for the data-types, in a perfect world yes, but there's so much variation in how wrds handles them that I end up checking everything in R anyways. Are datetimes datetimes or strings? Are permnos/gvkeys int/double/string or numeric/char? Since I can't (don't want to) remember which columns are properly typed, i just assume they are all numeric/char.

**Comment by iangow** [comment]
The `col_types` argument works for `wrds_to_parquet()` too. Lots of sample code [here](https://github.com/iangow/wrds_pg/blob/master/crsp/update_crsp.py). Once you've figured out the correct type, you never have to think about it again.

I think `TIMESTAMP WITH TIME ZONE` works best in PostgreSQL. Everything roundtrips nicely with R.

I would've guessed WRDS would be US/Eastern (I would assume as much for TAQ). 

https://github.com/r-dbi/RPostgres/issues/229

**Comment by EvanJo** [comment]
Stab taken. It works on linux/mac. I avoided things that depend on the user's local OS (local timezones etc), so it should also work on windows (but not tested).

**Comment by iangow** [comment]
> I can take a stab. mtime is in epoch time so that should be fine. The sas meta-data seems to be on central time.
> 
> Rather than converting to a time stamp, which could get messed up depending on the user's locale settings, I think it'd be best to just convert the sas time stamp to the proper UTC time.

I probably won't touch this over the next few weeks. I think (something like) `wrds_update_csv()` might be the thing to make.

**Comment by EvanJo** [comment]
annnnnd TAQ is GMT -5, WITHOUT daylight savings. of course.

I think i'm happy with wrds_update_csv() as is. Call it the WRDS timezone... since the date time goes from the sas metadata straight to the mtime, even those that are wrong are consistently wrong. It's good enough for the conditional updates i think.

alternative would be to get the mtime straight from the sas file (from proc contents). then everything is in epoch.

**Comment by iangow** [comment]
"the sas file (from proc contents)" is what I use for `wrds_update()` and `wrds_to_parquet()`, but for that purpose, it's literally a string. So whatever works to confirm it's the same file makes sense (I assume that's what we're talking about … I may have drifted into talking about datetimes in fields, which is a completely different topic).

**Comment by EvanJo** [comment]
Yes. Proc contents returns something like "Filename                    /wrds/nyse/sasdata/taqms/cq/cqm_20220701.sas7bdat"

you can then grab the epoch ctime with
`date -r /wrdslin/nyse/sasdata/taqms/cq/cqm_20220701.sas7bdat +%s
`

I'm just realizing that there used to be a symlink from /wrds --> /wrdslin . it's not available on my new wrds account anymore. weird.

**Comment by iangow** [comment]
https://github.com/iangow/wrds2pg/blob/4550f287e219035f2036b67acb73ebc9a9f60bee/wrds2pg/wrds2pg.py#L398-L401

**Comment by iangow** [comment]
I then stick this in the "comment" for the table in PostgreSQL and in the metadata for the parquet file. Only purpose is to indicate if a file has been changed (and therefore needs to be updated). I would say if `mtime` can store such information (robust to time zone, etc.), then it's good way to do it.

```
 Schema |       Name        | Type  | Owner | Persistence | Access method |    Size    |                          Description                           
--------+-------------------+-------+-------+-------------+---------------+------------+----------------------------------------------------------------
 comp   | aco_pnfnda        | table | comp  | permanent   | heap          | 35 MB      | Last modified: 10/30/2023 02:26:37
 comp   | adsprate          | table | comp  | permanent   | heap          | 130 MB     | Last modified: 10/30/2023 02:26:39
 comp   | anncomp           | table | comp  | permanent   | heap          | 265 MB     | Last modified: 10/04/2023 11:01:00
 comp   | ciks              | table | comp  | permanent   | heap          | 2632 kB    | Created using update_comp.py ON 2023-10-30 18:27:00
 comp   | co_adesind        | table | comp  | permanent   | heap          | 128 MB     | Last modified: 10/30/2023 02:24:24
 comp   | co_afnd2          | table | comp  | permanent   | heap          | 476 MB     | Last modified: 10/30/2023 02:24:36
 comp   | co_filedate       | table | comp  | permanent   | heap          | 132 MB     | Last modified: 10/30/2023 02:24:53
 comp   | co_hgic           | table | comp  | permanent   | heap          | 3240 kB    | Last modified: 10/30/2023 02:26:26
 comp   | co_ifndq          | table | comp  | permanent   | heap          | 1830 MB    | Last modified: 10/30/2023 02:25:38
 comp   | company           | table | comp  | permanent   | heap          | 24 MB      | Last modified: 10/30/2023 02:26:21
 comp   | funda             | table | comp  | permanent   | heap          | 1378 MB    | Last modified: 10/30/2023 02:38:21
 comp   | funda_fncd        | table | comp  | permanent   | heap          | 209 MB     | Last modified: 10/30/2023 02:39:46
 comp   | fundq             | table | comp  | permanent   | heap          | 3030 MB    | Last modified: 10/30/2023 02:46:36
 comp   | fundq_fncd        | table | comp  | permanent   | heap          | 635 MB     | Last modified: 10/30/2023 02:49:14
 comp   | g_chars           | table | comp  | permanent   | heap          | 8192 bytes | Last modified: 10/30/2023 12:00:33
 comp   | g_company         | table | comp  | permanent   | heap          | 62 MB      | Last modified: 10/30/2023 10:53:05
```

**Comment by EvanJo** [comment]
mtime is pretty handy since its in UTC epoch seconds, no need to worry about time zones or formatting. I'm a bit more worried about proc contents not returning consistent paths. I'm not sure why some wrds account have the symlink from /wrds to /wrdslin, and wonder if there might be other inconsistencies on the wrds server side. Any experience with path issues on wrds?

**Comment by iangow** [comment]
No experience with path issues. I think the "last modified" field in the SAS file might be more robust.

``` python
>>> from wrds2pg import get_modified_str
>>> get_modified_str("dsf", "crsp", wrds_id="iangow")
'Last modified: 01/17/2023 18:28:41'
```

I wrote parts of this code more than ten years ago (originally in Perl and before I used Git), so sometimes I don't remember any details. But I think the logic of `get_modified_str()` has stood the test of time.

I would view the problem as one of encoding `'Last modified: 01/17/2023 18:28:41'` and storing it somewhere in the file. It seems `mtime` would be the best place. It seems the `Last modified: ` part is invariant (see code above). So it's really just converting `01/17/2023 18:28:41` to something that can be stored as `mtime`. That conversion could be to any time zone, so long as the encoding and decoding are consistent. (If `mtime` is stored by file systems as epoch time, then I guess there are no time zone issues … I figure we can consider potential issues with [leap seconds](https://www.ntp.org/reflib/leap/) as out of scope!)

**Comment by EvanJo** [comment]
Then my existing pull request would be exactly that, since it converts the sas last modified to epoch and stores it in mtime.  Since it always reads the sas data last modified, and writes it to epoch, it will always be self-consistent for the update purpose. 

The epoch time is just sometimes "wrong", since there's no way of telling what time zone the sas data last modified time is in. I hand check a few (comparing the sas last modified to the mtime on wrds), and it looks like optionm is in central with DST and taq, as you guessed, is in Eastern without daylight savings. 

None of this affects the code. Just some tables would be off by an hour occasionally if someone compares the mtime to the wrds server mtime.

**Comment by iangow** [comment]
I worked through your pull request. There were some conflicts and I cleaned some things up a bit. So I will probably implement code based on this (which is based on your code). I mainly just to check that some arguments are relevant (e.g., `sas_schema`).

https://github.com/iangow/wrds2pg/blob/8602092b9807b810330588d76d2aa46ba675a48d/last_modified_csv.ipynb

**Comment by EvanJo** [comment]
Ah, good point. i kept all the same arguments so that I can easily (lazily) switch between the two.

**Comment by iangow** [comment]
Still work to be done, but 2cfab1b addresses much of this.

**Comment by iangow** [comment]
@EvanJo I did a bunch of work on this yesterday (was a "chaperone" at an event for my daughters' school, so didn't have much to do). But I mistakenly pushed this work-in-progress stuff to `master` (I thought I'd made the commits to a `dev` branch). So far it seems to be all working, but now is perhaps not the best time to reinstall from `master`.

**Comment by iangow** [comment]
This is basically done. There are three main functions:

- `wrds_update()`: Updates PostgreSQL using WRDS SAS. [Keys off](https://en.wiktionary.org/wiki/key_off) usual PostgreSQL environmental variables (`PGHOST`, `PGDATABASE`, etc.).
- `wrds_update_csv()`: Updates gzipped CSV repository using WRDS SAS. Keys off `CSV_DIR`.
- `wrds_update_pq()`: Updates parquet repository using WRDS SAS. Keys off `DATA_DIR`. (This function replaces `wrds_to_parquet()`, which is no more.)
